File: mil/sdk-ts/src/crypto.ts
// MIL data-plane crypto (design §3.2): ML-KEM-1024 encapsulation,
// HKDF-SHA3-512 key schedule, and direction-keyed AES-256-GCM framing.
// Byte-compatible with `misaka_mil_channel::{kem, session, secure}`.

import { ml_kem1024 } from "@noble/post-quantum/ml-kem";
import { hkdf } from "@noble/hashes/hkdf";
import { sha3_512 } from "@noble/hashes/sha3";
import { gcm } from "@noble/ciphers/aes";
import { DOMAINS } from "./hash.ts";

const enc = new TextEncoder();

export const KEM_CT_LEN = 1568;
export const KEM_EK_LEN = 1568;

/** Requester-side ML-KEM-1024 encapsulation to a provider's pk_kem. */
export function encapsulate(pkKem: Uint8Array): { cipherText: Uint8Array; sharedSecret: Uint8Array } {
  const { cipherText, sharedSecret } = ml_kem1024.encapsulate(pkKem);
  return { cipherText, sharedSecret };
}

export const Direction = { ClientToProvider: 0x01, ProviderToClient: 0x02 } as const;
export type Direction = (typeof Direction)[keyof typeof Direction];

/** HKDF-SHA3-512 direction keys, info = "misaka-mil-v1/kdf" ‖ session_id.
 *  Salt = None (RFC5869 → HashLen zeros; matches the Rust `Hkdf::new(None, ss)`). */
export function deriveSessionKeys(sharedSecret: Uint8Array, sessionIdV: Uint8Array): { kC2P: Uint8Array; kP2C: Uint8Array } {
  const info = new Uint8Array(DOMAINS.KDF.length + sessionIdV.length);
  info.set(enc.encode(DOMAINS.KDF));
  info.set(sessionIdV, DOMAINS.KDF.length);
  const okm = hkdf(sha3_512, sharedSecret, undefined, info, 64);
  return { kC2P: okm.slice(0, 32), kP2C: okm.slice(32, 64) };
}

function nonceFor(direction: Direction, seq: bigint): Uint8Array {
  const n = new Uint8Array(12);
  n[0] = direction;
  new DataView(n.buffer).setBigUint64(4, seq, true);
  return n;
}

function aadFor(sessionIdV: Uint8Array, direction: Direction, frameType: number, seq: bigint): Uint8Array {
  const aad = new Uint8Array(74);
  aad.set(sessionIdV.subarray(0, 64), 0);
  aad[64] = direction;
  aad[65] = frameType & 0xff;
  new DataView(aad.buffer).setBigUint64(66, seq, true);
  return aad;
}

/** Sealing half: owns a direction key and a monotonic send counter. */
export class SendCipher {
  private nextSeq = 0n;
  private key: Uint8Array;
  private sessionIdV: Uint8Array;
  private direction: Direction;
  constructor(key: Uint8Array, sessionIdV: Uint8Array, direction: Direction) {
    this.key = key;
    this.sessionIdV = sessionIdV;
    this.direction = direction;
  }

  seal(frameType: number, plaintext: Uint8Array): { seq: bigint; ciphertext: Uint8Array } {
    const seq = this.nextSeq;
    this.nextSeq += 1n;
    const aad = aadFor(this.sessionIdV, this.direction, frameType, seq);
    const ciphertext = gcm(this.key, nonceFor(this.direction, seq), aad).encrypt(plaintext);
    return { seq, ciphertext };
  }
}

/** Opening half: enforces strict in-order sequence numbers. */
export class RecvCipher {
  private expectedSeq = 0n;
  private key: Uint8Array;
  private sessionIdV: Uint8Array;
  private direction: Direction;
  constructor(key: Uint8Array, sessionIdV: Uint8Array, direction: Direction) {
    this.key = key;
    this.sessionIdV = sessionIdV;
    this.direction = direction;
  }

  open(frameType: number, seq: bigint, ciphertext: Uint8Array): Uint8Array {
    if (seq !== this.expectedSeq) {
      throw new Error(`MIL record out of order: expected ${this.expectedSeq}, got ${seq}`);
    }
    const aad = aadFor(this.sessionIdV, this.direction, frameType, seq);
    const plaintext = gcm(this.key, nonceFor(this.direction, seq), aad).decrypt(ciphertext);
    this.expectedSeq += 1n;
    return plaintext;
  }
}

File: mil/sdk-swift/Sources/MilSDK/Crypto.swift
import Foundation

/// The PQ + AEAD primitives the MIL data plane needs (design §3.2). These are
/// abstracted so the pure-Swift parts (Borsh, Hash64, receipt layout) build and
/// test with no external crypto dependency; a platform integration binds this
/// protocol to CryptoKit (macOS 15+ / iOS 18+ ship `MLKEM1024` and `MLDSA`) or
/// swift-crypto.
///
/// Byte contracts (all must match `misaka_mil_channel`):
/// - `encapsulate`: ML-KEM-1024, returns (ciphertext[1568], sharedSecret[32]).
/// - `hkdfSha3_512`: HKDF with SHA3-512, salt = nil (RFC5869 → HashLen zeros).
/// - `aesGcmSeal`/`aesGcmOpen`: AES-256-GCM, 12-byte nonce, ciphertext‖tag(16).
/// - `mldsa87Verify`: ML-DSA-87 verify with a context string (FIPS 204).
public protocol MilCryptoProvider {
    func encapsulate(pkKem: [UInt8]) throws -> (ciphertext: [UInt8], sharedSecret: [UInt8])
    func hkdfSha3_512(ikm: [UInt8], info: [UInt8], length: Int) -> [UInt8]
    func aesGcmSeal(key: [UInt8], nonce: [UInt8], aad: [UInt8], plaintext: [UInt8]) -> [UInt8]
    func aesGcmOpen(key: [UInt8], nonce: [UInt8], aad: [UInt8], ciphertext: [UInt8]) throws -> [UInt8]
    func mldsa87Verify(publicKey: [UInt8], message: [UInt8], signature: [UInt8], context: [UInt8]) -> Bool
}

public enum Direction: UInt8 { case clientToProvider = 0x01, providerToClient = 0x02 }

/// Derive the two direction keys: `info = "misaka-mil-v1/kdf" ‖ session_id`,
/// first 32 bytes = client→provider, last 32 = provider→client.
public func deriveSessionKeys(_ crypto: MilCryptoProvider, sharedSecret: [UInt8], sessionId: [UInt8])
    -> (kC2P: [UInt8], kP2C: [UInt8])
{
    let okm = crypto.hkdfSha3_512(ikm: sharedSecret, info: MilHash.Domain.kdf + sessionId, length: 64)
    return (Array(okm[0..<32]), Array(okm[32..<64]))
}

public func aeadNonce(direction: Direction, seq: UInt64) -> [UInt8] {
    var n = [UInt8](repeating: 0, count: 12)
    n[0] = direction.rawValue
    for i in 0..<8 { n[4 + i] = UInt8((seq >> (8 * UInt64(i))) & 0xff) }
    return n
}

public func aeadAad(sessionId: [UInt8], direction: Direction, frameType: UInt8, seq: UInt64) -> [UInt8] {
    var aad = [UInt8](repeating: 0, count: 74)
    for i in 0..<64 { aad[i] = sessionId[i] }
    aad[64] = direction.rawValue
    aad[65] = frameType
    for i in 0..<8 { aad[66 + i] = UInt8((seq >> (8 * UInt64(i))) & 0xff) }
    return aad
}

/// Sealing half with a monotonic send counter.
public struct SendCipher {
    private let crypto: MilCryptoProvider
    private let key: [UInt8]
    private let sessionId: [UInt8]
    private let direction: Direction
    private var nextSeq: UInt64 = 0

    public init(crypto: MilCryptoProvider, key: [UInt8], sessionId: [UInt8], direction: Direction) {
        self.crypto = crypto; self.key = key; self.sessionId = sessionId; self.direction = direction
    }

    public mutating func seal(frameType: UInt8, plaintext: [UInt8]) -> (seq: UInt64, ciphertext: [UInt8]) {
        let seq = nextSeq
        nextSeq += 1
        let ct = crypto.aesGcmSeal(
            key: key, nonce: aeadNonce(direction: direction, seq: seq),
            aad: aeadAad(sessionId: sessionId, direction: direction, frameType: frameType, seq: seq),
            plaintext: plaintext)
        return (seq, ct)
    }
}

/// Opening half enforcing strict in-order sequence numbers.
public struct RecvCipher {
    private let crypto: MilCryptoProvider
    private let key: [UInt8]
    private let sessionId: [UInt8]
    private let direction: Direction
    private var expectedSeq: UInt64 = 0

    public init(crypto: MilCryptoProvider, key: [UInt8], sessionId: [UInt8], direction: Direction) {
        self.crypto = crypto; self.key = key; self.sessionId = sessionId; self.direction = direction
    }

    public mutating func open(frameType: UInt8, seq: UInt64, ciphertext: [UInt8]) throws -> [UInt8] {
        guard seq == expectedSeq else {
            throw MilError.recordOutOfOrder(expected: expectedSeq, got: seq)
        }
        let pt = try crypto.aesGcmOpen(
            key: key, nonce: aeadNonce(direction: direction, seq: seq),
            aad: aeadAad(sessionId: sessionId, direction: direction, frameType: frameType, seq: seq),
            ciphertext: ciphertext)
        expectedSeq += 1
        return pt
    }
}

public enum MilError: Error {
    case recordOutOfOrder(expected: UInt64, got: UInt64)
    case versionMismatch(UInt16)
    case attestationRejected(String)
    case receiptInvalid(String)
    case transcriptMismatch
}

/// Verify a receipt's ML-DSA-87 signature under the MIL receipt context.
public func verifyReceipt(_ crypto: MilCryptoProvider, _ r: SignedReceipt) -> Bool {
    guard r.body.version == MilProtocol.version,
          r.providerPk.count == 2592, r.signature.count == 4627 else { return false }
    return crypto.mldsa87Verify(
        publicKey: r.providerPk, message: r.body.signingMessage(),
        signature: r.signature, context: MilHash.Domain.receiptCtx)
}
